### PR TITLE
Speedup visualization and widget responses

### PR DESCRIPTION
**Pull request by hubertp**
### Pull Request Description

`ExecuteJob` can now be interrupted.
We now have a separate threadpool for visualization-related jobs. 

### Important Notes

In a lock step situation, a job or command could have been interrupted while waiting for one of the locks. As locks ensured only that they were released once all of them have been acquired this could leave engine in a broken state.
Once `ExecuteJob` could be interrupted this became a blocker as it prevented project startup almost in every case.
The change also makes it careful to avoid constant `ExecuteJob` restarts.

Addresses #9278. There will be follow up work.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by hubertp** [comment]
After experimenting a bit with interruptions I'm afraid this won't cut it. If scrolling entries in a component browser sends a single (or,  as we see in real life, 2) request(s) on every update, and every update triggers a cancellation of an existing execution, then never finishing.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/ThreadExecutors.java [hunk 1]
```diff
@@ -20,13 +18,28 @@ final class ThreadExecutors {
...
   }
 
-  final ExecutorService newCachedThreadPool(String name, boolean systemThread) {
+  ExecutorService newCachedThreadPool(String name, boolean systemThread) {
     var s = Executors.newCachedThreadPool(new Factory(name, systemThread));
     pools.put(s, name);
     return s;
   }
 
-  final ExecutorService newFixedThreadPool(int cnt, String name, boolean systemThread) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/ThreadExecutors.java:29]: Why do we remove `final`? We don't want anyone to ever override these methods.

**Review by JaroslavTulach — COMMENTED** [review]
I see a lot of Scala changes in the language server, very few changes in the engine and I don't have an answer to an important question: **how do we actually interrupt the execution?**

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/ThreadExecutors.java (hunk 1 above)
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/ThreadExecutors.java:29]: Class is already final. IDE suggested that

**Comment by hubertp** [comment]
> I see a lot of Scala changes in the language server, very few changes in the engine and I don't have an answer to an important question: **how do we actually interrupt the execution?**

`abortJobs` already "did abort" executions. Except the `ExecuteJob` was cancellable but not interruptible. Now it is both.

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ExecuteJob.scala [hunk 2]
```diff
@@ -24,65 +24,72 @@ class ExecuteJob(
       isCancellable = true,
       // TODO[MK]: make this interruptible when https://github.com/oracle/graal/issues/3590
       // is resolved
-      mayInterruptIfRunning = false
+      mayInterruptIfRunning = true
```
Comment (4e6) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ExecuteJob.scala:27]: Can we make it interruptible even if the issue in the comment is not resolved?

Comment (4e6) [reply engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ExecuteJob.scala:27]: I wonder, if we make all the execute jobs interruptible, would it break the scenario with the `executeExpression` spam, when the user quickly switches between two entries in the component browser menu?

**Review by 4e6 — APPROVED** [review]
Just some concerns about making all execute jobs interruptible. Otherwise it looks good.

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ExecuteJob.scala (hunk 2 above)
Comment (hubertp) [reply engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ExecuteJob.scala:27]: > Can we make it interruptible even if the issue in the comment is not resolved?

I tried to reproduce but can't anymore. I will modify the comment

Comment (hubertp) [reply engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ExecuteJob.scala:27]: > I wonder, if we make all the execute jobs interruptible, would it break the scenario with the `executeExpression` spam, when the user quickly switches between two entries in the component browser menu?

That is actually a good point. Looks like I won't get away from introducing a new kind of `ExecuteJob` after all

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/ThreadExecutors.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/ThreadExecutors.java:29]: I'd suggest to disable this rule in your IDE. There is nothing wrong on making sure _methods aren't virtual_ multiple times.

**Comment by JaroslavTulach** [comment]
> `abortJobs` already "did abort" executions. Except the `ExecuteJob` was cancellable but not interruptible. Now it is both.

I am trying to understand whether do we use the _right technique_ when interrupting on the Truffle level? However I have so far been unable to find out what do we actually do interrupt an execution?

> > I see a lot of Scala changes in the language server, very few changes in the engine and I don't have an answer to an important question: **how do we actually interrupt the execution?**

If you believe that question isn't relevant, then continue without my review.

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ExecuteJob.scala (hunk 2 above)
Comment (hubertp) [reply engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/ExecuteJob.scala:27]: > > I wonder, if we make all the execute jobs interruptible, would it break the scenario with the `executeExpression` spam, when the user quickly switches between two entries in the component browser menu?
> 
> That is actually a good point. Looks like I won't get away from introducing a new kind of `ExecuteJob` after all

This has been addressed. Spamming `executeExpression` won't flood the engine with multiple `ExecuteJob`s.

**Comment by hubertp** [comment]
> > `abortJobs` already "did abort" executions. Except the `ExecuteJob` was cancellable but not interruptible. Now it is both.
> 
> I am trying to understand whether do we use the _right technique_ when interrupting on the Truffle level? However I have so far been unable to find out what do we actually do interrupt an execution?
> 
> > > I see a lot of Scala changes in the language server, very few changes in the engine and I don't have an answer to an important question: **how do we actually interrupt the execution?**
> 
> If you believe that question isn't relevant, then continue without my review.

TBH I'm not sure how _relevant_ it is because I don't fully grasp it. This PR does not change the way we interrupt jobs. It's still using `Future.cancel` for that. I think it would be best to have a meeting about it to clarify your concerns.
Either way, this PR already improves the status quo in the presence of for example multiple edits (previously execution would wait until the future finishes). And `executeExpression` only cancels related jobs, not all jobs. Not to mention the fact that locking in the presence of interruptions was simply broken.

**Pull request merged**